### PR TITLE
[QOL-5710] restrict user_show API to organisation admins or self

### PR DESCRIPTION
**Pull request by ThrawnCA**

### CODE REVIEW

**Comment by ThrawnCA** [comment]
> Can we add some selenium tests for this to the circleci pipeline.

This extension doesn't currently have a CircleCI pipeline. I can raise a related JIRA issue for that if you want.

**Comment by duttonw** [comment]
ckan_1      |   File "/app/ckanext/qgov/common/plugin.py", line 27, in <module>
ckan_1      |     import ckanext.qgov.common.intercepts as intercepts
ckan_1      |   File "/app/ckanext/qgov/common/intercepts.py", line 25, in <module>
ckan_1      |     from ckanext.qgov.common.controller import plugin

**Comment by duttonw** [comment]
Thanks for adding the cicd pipeline, looks like we caught a couple of bugs because of it :)

File: .docker/scripts/init-ext.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+set -e
+
+. /app/ckan/default/bin/activate
+
+pip install -r "/app/requirements.txt"
+pip install -r "/app/requirements-dev.txt"
+python setup.py develop
+
+# Validate that the extension was installed correctly.
+if ! pip list | grep ckanext-qgov > /dev/null; then echo "Unable to find the extension in the list"; exit 1; fi
```
Comment (duttonw) [inline .docker/scripts/init-ext.sh]: For another time, we should look at breaking out the plugin names so these scripts are even more generic.

File: ckanext/qgov/common/plugin.py [hunk 2]
```diff
@@ -130,12 +130,34 @@ def related_update(context, data_dict):
...
+        user_obj = model.User.get(id)
+        if user_obj:
+            return {'success': requester == user_obj.name}
+
+    return {'success': False}
+
+
+def _requester_is_admin(context):
+    requester = context.get('user')
+    return authz.has_user_permission_for_some_org(requester, 'update')
```
Comment (duttonw) [inline ckanext/qgov/common/plugin.py]: comment on this function please, how does checking for 'update' mean they are admin?

Comment (ThrawnCA) [reply ckanext/qgov/common/plugin.py]: That's what organisation admin privileges look like. See `ckan/logic/auth/update.py`, `group_edit_permissions` function.

File: .docker/scripts/init-ext.sh (hunk 1 above)
Comment (ThrawnCA) [reply .docker/scripts/init-ext.sh]: Done.

File: test/features/steps/steps.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+
+@step('I go to register page')
+def go_to_register_page(context):
+    when_i_visit_url(context, '/user/register')
+
+
+@step('I go to the user autocomplete API')
+def go_to_user_autocomplete(context):
+    when_i_visit_url(context, '/api/2/util/user/autocomplete?q=admin')
```
Comment (duttonw) [inline test/features/steps/steps.py]: should take variable for query or update step to 'I go the user autocomplete API and search admin

File: test/features/steps/steps.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+
+@step('I go to the user list API')
+def go_to_user_list(context):
+    when_i_visit_url(context, '/api/3/action/user_list')
+
+
+@step('I view the "{group_id}" group API including users')
```
Comment (duttonw) [inline test/features/steps/steps.py]: make it take show user flag so you can do the inverse test to see if non standard users can still work.

I view the "{group_id}" group API including users "{boolUsersView}"

I view the "department-of-health" group API including users "true"
I view the "department-of-health" group API including users "false"

File: test/features/steps/steps.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+def go_to_group_including_users(context, group_id):
+    when_i_visit_url(context, r'/api/3/action/group_show?id={}&include_users=true'.format(group_id))
+
+
+@step('I view the "{organisation_id}" organisation API including users')
```
Comment (duttonw) [inline test/features/steps/steps.py]: same as comment on line 51

File: test/features/users.feature [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        Given "Foodie" as the persona
+        When I log in
+        And I view the "department-of-health" organisation API including users
+        And I take a screenshot
+        Then I should see an element with xpath "//*[contains(string(), '"success": false,') and contains(string(), 'Authorization Error')]"
+
+    Scenario: Ensure organisation membership is not accessible anonymously
+        When I view the "department-of-health" organisation API including users
+        And I take a screenshot
+        Then I should see an element with xpath "//*[contains(string(), '"success": false,') and contains(string(), 'Authorization Error')]"
```
Comment (duttonw) [inline test/features/users.feature]: add positive tests for viewing organisation api or group api with user viewing false for all personas that it works (including anon)

Comment (ThrawnCA) [reply test/features/users.feature]: Huh, that actually caught a legitimate bug. It seems that custom auth functions need to be annotated to permit anonymous access.

**Pull request merged**